src/test/librbd/mock/MockOperations.h

```diff
@@ -25,7 +25,7 @@ struct MockOperations {
                                     ProgressContext &prog_ctx,
                                     Context *on_finish,
                                     uint64_t journal_op_tid));
-  MOCK_METHOD5(snap_create, void(const cls::rbd::SnapshotNamespace &snapshot_namespace,
+  MOCK_METHOD5(snap_create, void(cls::rbd::MirrorSnapshotNamespace &snapshot_namespace,
```

Comment:
This seems wrong.  `snap_create` is one of the generic mock methods, why is it being constrained to mirror snapshots?

Reply:
It doesn't compile either:
```
[2229/2624] Building CXX object src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_sync/test_mock_SyncPointCreateRequest.cc.o
FAILED: src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_sync/test_mock_SyncPointCreateRequest.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -isystem /opt/ceph/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googlemock/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googlemock -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googletest/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googletest -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/src/pmdk/src/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/sdk/include -g -Werror -fPIE -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++20 -MD -MT src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_sync/test_mock_SyncPointCreateRequest.cc.o -MF src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_sync/test_mock_SyncPointCreateRequest.cc.o.d -o src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_sync/test_mock_SyncPointCreateRequest.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rbd_mirror/image_sync/test_mock_SyncPointCreateRequest.cc
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rbd_mirror/image_sync/test_mock_SyncPointCreateRequest.cc:26:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:116:5: error: non-const lvalue reference to type 'cls::rbd::MirrorSnapshotNamespace' cannot bind to a temporary of type 'cls::rbd::UserSnapshotNamespace'
    cls::rbd::UserSnapshotNamespace(), sync_point.snap_name.c_str(),
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:103:3: note: in instantiation of member function 'rbd::mirror::image_sync::SyncPointCreateRequest<librbd::(anonymous namespace)::MockTestImageCtx>::send_create_snap' requested here
  send_create_snap();
  ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:88:60: note: in instantiation of member function 'rbd::mirror::image_sync::SyncPointCreateRequest<librbd::(anonymous namespace)::MockTestImageCtx>::handle_refresh_image' requested here
    SyncPointCreateRequest<I>, &SyncPointCreateRequest<I>::handle_refresh_image>(
                                                           ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:80:3: note: in instantiation of member function 'rbd::mirror::image_sync::SyncPointCreateRequest<librbd::(anonymous namespace)::MockTestImageCtx>::send_refresh_image' requested here
  send_refresh_image();
  ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:64:33: note: in instantiation of member function 'rbd::mirror::image_sync::SyncPointCreateRequest<librbd::(anonymous namespace)::MockTestImageCtx>::handle_update_sync_points' requested here
    &SyncPointCreateRequest<I>::handle_update_sync_points>(this);
                                ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:50:3: note: in instantiation of member function 'rbd::mirror::image_sync::SyncPointCreateRequest<librbd::(anonymous namespace)::MockTestImageCtx>::send_update_sync_points' requested here
  send_update_sync_points();
  ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rbd_mirror/image_sync/test_mock_SyncPointCreateRequest.cc:112:8: note: in instantiation of member function 'rbd::mirror::image_sync::SyncPointCreateRequest<librbd::(anonymous namespace)::MockTestImageCtx>::send' requested here
  req->send();
       ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/mock/MockOperations.h:28:3: note: passing argument to parameter 'gmock_a0' here
  MOCK_METHOD5(snap_create, void(cls::rbd::MirrorSnapshotNamespace &snapshot_namespace,
  ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googlemock/include/gmock/gmock-function-mocker.h:322:30: note: expanded from macro 'MOCK_METHOD5'
#define MOCK_METHOD5(m, ...) GMOCK_INTERNAL_MOCK_METHODN(, , m, 5, __VA_ARGS__)
                             ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googlemock/include/gmock/gmock-function-mocker.h:472:3: note: expanded from macro 'GMOCK_INTERNAL_MOCK_METHODN'
  GMOCK_INTERNAL_MOCK_METHOD_IMPL(                                        \
  ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/googletest/googlemock/include/gmock/gmock-function-mocker.h:144:35: note: expanded from macro 'GMOCK_INTERNAL_MOCK_METHOD_IMPL'
      _MethodName(GMOCK_PP_REPEAT(GMOCK_INTERNAL_PARAMETER, _Signature, _N))   \
                                  ^
1 error generated.
```

Reply:
> This seems wrong. snap_create is one of the generic mock methods, why is it being constrained to mirror snapshots?

It would be not possible to use it within expect_create_snapshot()

For example

```
                  Invoke([this, &mock_image_ctx, r](
                             const cls::rbd::SnapshotNamespace &ns,
                             const std::string& snap_name,
                             uint64_t flags,
                             ProgressContext &prog_ctx,
                             Context *on_finish) {
                           if (r != 0) {
                             return;
                           }
                           snap_create(mock_image_ctx, ns, snap_name);
                         }),
```

`ns.complete` is absent in the above case as `ns` is of type `const cls::rbd::SnapshotNamespace`  i.e its `const` and not of type ` cls::rbd::MirrorSnapshotNamespace`

I was able to compile it and execute the tests, not sure why it was not breaking compilation for me.

Is it ok if I declare a new `MOCK_METHOD6(snap_create, void(cls::rbd::MirrorSnapshotNamespace &snapshot_namespace, ...)` instead of editing `MOCK_METHOD5(snap_create, void(cls::rbd::SnapshotNamespace &snapshot_namespace, ...)` ?

Reply:
> `ns.complete` is absent in the above case as `ns` is of type `const cls::rbd::SnapshotNamespace` i.e its `const` and not of type ` cls::rbd::MirrorSnapshotNamespace`

`cls::rbd::SnapshotNamespace` is a variant that can hold `cls::rbd::MirrorSnapshotNamespace` as one of the alternatives:
```
typedef std::variant<UserSnapshotNamespace,
		     GroupSnapshotNamespace,
		     TrashSnapshotNamespace,
		     MirrorSnapshotNamespace,
		     UnknownSnapshotNamespace> SnapshotNamespaceVariant;

struct SnapshotNamespace : public SnapshotNamespaceVariant {
```
See https://en.cppreference.com/w/cpp/utility/variant, in particular `std::get` and `std::get_if` methods.  One example of how it's used is just a few lines above the line that you are modifying in this PR ;)

Reply:
> Is it ok if I declare a new `MOCK_METHOD6(snap_create, void(cls::rbd::MirrorSnapshotNamespace &snapshot_namespace, ...)` instead of editing `MOCK_METHOD5(snap_create, void(cls::rbd::SnapshotNamespace &snapshot_namespace, ...)` ?

No, because the real `snap_create` method also takes `cls::rbd::SnapshotNamespace`.

That said, I don't understand why setting an expectation on snapshot creation is desired.  It seems like incomplete snapshot creation could be part of the test setup and then the expectation could be just that `CreatePrimaryRequest` invokes `UnlinkPeerRequest` for a respective snapshot?